core/include/mutex.h

```diff
@@ -287,14 +265,12 @@ static inline int mutex_trylock(mutex_t *mutex)
  *
  * @post    The mutex @p is locked and held by the calling thread.
  */
-#if (MAXTHREADS > 1) || DOXYGEN
-void mutex_lock(mutex_t *mutex);
-#else
-/**
- * @brief   dummy implementation for when no scheduler is used
- */
 static inline void mutex_lock(mutex_t *mutex)
 {
+#if (MAXTHREADS > 1)
+    mutex_lock_internal(mutex, true);
+#else
+    /* dummy implementation for when no scheduler is used */
     /* (ab)use next pointer as lock variable */
     volatile uintptr_t *lock = (void *)&mutex->queue.next;
```

Comment:
I'm unsure whether `volatile` has the right semantics for this (it does guarantee that the write is made, but does it ensure the ordering relative to other operations on memory?), but that's not a thing that needs to be addressed here.

Reply:
this is indeed another misuse of `volatile`. Previously they seemed to "fix" things because `irq_disable()` and `irq_restore()` did not include compiler memory barriers, but the corresponding inline asm was also marked as `volatile`. I am quite sure that the actual bugs are fixed for all platforms now, so that this workaround should be dropped now. But that is something for a different PR IMO